extra_data/copy.py

```diff
@@ -0,0 +1,182 @@
+import sys
+from argparse import ArgumentParser
+from pathlib import Path
+from typing import Union
+
+import h5py
+
+from .utils import progress_bar
+
+__all__ = ["copy_structure"]
+
+
+def progress(processed, total, *, show=True):
+    """Show progress information"""
+    if not show:
+        return
+
+    pbar = progress_bar(processed, total)
+    if sys.stderr.isatty():
+        # "\x1b[2K": delete whole line, "\x1b[1A": move up cursor
+        print("\x1b[2K\x1b[1A\x1b[2K\x1b[1A", file=sys.stderr)
+    print(pbar, file=sys.stderr)
+
+
+class Cloner:
+    def __init__(self, input, output, *, run_data=False, control_data=False):
+        self.run_data = run_data
+        self.control_data = control_data
+        self.visited = {}
+
+        if output.file.mode == "r":
+            raise ValueError("Output file must be writeable.")
+        self.visit(input, output)
+
+    @staticmethod
+    def _copy_attrs(input, output):
+        for key, value in input.attrs.items():
+            output.attrs.create(key, value)
+
+    def visit(self, obj, output):
+        if obj.name != "/":
+            link = obj.file.get(obj.name, getlink=True)
+            if isinstance(link, h5py.SoftLink):
+                output[obj.name] = h5py.SoftLink(link.path)
+                return
+            elif isinstance(link, h5py.ExternalLink):
+                # TODO do we want to support external links?
+                # this *might* work, but external softlinks may point to non reacheable data
+                # with h5py.File(link.filename) as ext:
+                #     Cloner(ext[link.path], output[obj.name], run_data=self.run_data, control_data=self.control_data)
```

Comment:
Since we've got the code here ready, why not just copy the link itself like we're doing for soft links? I think just doing `output[obj.name] = link` should do it.

Reply:
good point, although, the idea is that these generated files might be used somewhere with no access the external files it links to (particularly test envs). But sure I'll add support to it.

Reply:
If there's a voview file inside the run directory, it will contain links to other files in the same directory. Those should be done as relative paths, so the copy of that file should point to the other files in the same directory as itself. :crossed_fingers: